### ISSUE TITLE
Odd results for occurrences of a macro

### ISSUE THREAD

**Issue by ckandeler** [clangd/clangd#1299]
Consider this piece of code:
```
struct Foo {
    int _x;
    int _Y;

#define FOO_XLIST(F) F(x) F(Y)

#define FOO_GETTER(field) int field() const { return _##field; }
FOO_XLIST(FOO_GETTER)
#undef FOO_GETTER
};
```
Observation 1: documentHighlight for FOO_GETTER yields no results, regardless of which occurrence is tried.
Observation 2: references for FOO_GETTER returns the first and the third occurrence (i.e. the define and the undef), but not the second. Interestingly, this is even true when the second one is used as the position in the request, i.e. clangd seems to know it's the same symbol as the other two, but it's still not part of the result set.
Observation 3: Rather strangely, documentHighlight for the two struct members (x and _Y) also returns the second occurrence of FOO_GETTER. The same goes for references, leading to e.g. unwanted renaming effects.

**Comment by ckandeler** [comment]
>     * unwanted renaming effects: can you elaborate? The behavior I see seems reasonable here: we perform the rename but ignore the occurrence expanded from macros, because there's no safe/predictable edit here.

I see. In our client (Qt Creator), we do the renaming ourselves based on the found references, as we (optionally) do extra stuff like renaming files corresponding to classes, and also we allow users to opt out of the renaming for occurrences they don't want changed. If we could tell these "special" references apart, we could pre-unselect them.

**Comment by sam-mccall** [comment]
I think filtering out edits where the replaced text != the original selected identifier would go a long way - we do that for index results IIRC. I know it sounds like a hack...

There are some other differences:
 - canonicalize the symbol before performing find refs (e.g. specialization to primary)
 - add all overrides to the query (as of recently)
 - patch up index results if the code has changed since indexing

Any reason you can't do this on top of rename instead of references? Then you'd pick up these behaviors too.

**Comment by HighCommander4** [comment]
>    * references: formally the `FOO_GETTER` in `FOO_XLIST(FOO_GETTER)` is just a token, not a macro reference. (consider `#define FOO_XLIST(X) #X`). The macro-reference to `FOO_GETTER` occurs inside a macro expansion, and we'd need to do some work to tie it back to the code (and to FOO_GETTER rather than FOO_XLIST).

#346 is similar

**Comment by ckandeler** [comment]
> Any reason you can't do this on top of rename instead of references? Then you'd pick up these behaviors too.

TBH, I've never looked at clangd's rename functionality more closely because we already had our own (for historical reasons, plus we support different code model backends). Now that I realize that rename is not the same as find references + replace, I might have to reconsider. I'll investigate and report back. 
However, I do wonder why references can't report the same set of results that rename would change. For instance, the override thing mentioned above seems like useful information that one would like to see even without renaming.

**Comment by sam-mccall** [comment]
> I do wonder why references can't report the same set of results that rename would change

They're certainly similar operations but i don't see how we can make them identical without breaking things:
- Distinction between a name and a symbol. Given A::foo and B::foo both overriding Base::foo, renaming A::foo should rename B::foo and its callers (or we break the code) but find-refs on A::foo should not find calls to B::foo (we're typically trying to find callsites, and these are false positives).
- i think it's important that find-refs returns even references obscured by macros, as checking references exhaustively is a key way to reason about data flow. OTOH silently editing stuff in macro bodies is unreasonably dangerous.

**Comment by ckandeler** [comment]
Strangely, prepareRename succeeds for _x, but fails for _Y with:
    "Cannot rename symbol: symbol may be used in other files (not eligible for indexing)"
Not sure what the difference is supposed to be between these symbols. There is only one file.

**Comment by ckandeler** [comment]
Ah, because it's a reserved name. I suppose it makes sense in a way, but what if you mistakenly used a reserved name and now you want to fix it by renaming ...?

**Comment by sam-mccall** [comment]
Yeah, that seems a bit unfortunate. It falls out of other decisions. In this case we're not indexing reserved names at all, so renaming them wouldn't be safe.
The goals include hiding `_M_ptr`, `__vector_base` etc type symbols that leak out of stdlib implementations, reducing the size of the stdlib index that will soon be on by default, reducing the confusion when remote index & local build use different stdlib implementations...
(There are different ways of achieving these goals too, but this option was very simple and solved multiple problems)
